File: labs/lab1/python/load.py
import pandas as pd
from os.path import join as oj


def load_dates_data(path='data'):
    ''' 
    Load the dates data

    Parameters
    ----------
    path : str
        The path to the data directory

    Returns
    -------
    df : pd.DataFrame
        Dataframe containing the epoch information with three columns: 
        'epoch number', 'date', 'day'
    '''
    epoch_nums = pd.read_csv(
        oj(path, 'sonoma-dates-epochNums.txt'), sep=' ', header=None
    ).T
    epoch_nums.columns = ['number']

    epoch_dates = pd.read_csv(
        oj(path, 'sonoma-dates-epochDates.txt'), sep="'\\s+", header=None, engine='python'
    ).T
    epoch_dates.columns = ['date']
    epoch_dates['date'] = epoch_dates['date'].str.replace("'", "")
    
    epoch_days = pd.read_csv(
        oj(path, 'sonoma-dates-epochDays.txt'), sep=' ', header=None
    ).T
    epoch_days.columns = ['day']
    
    epoch_data = pd.concat([epoch_nums, epoch_dates, epoch_days], axis=1)
    return epoch_data


def load_redwood_data(path='data', source='all'):
    ''' 
    Load the redwood data

    Parameters
    ----------
    path : str
        The path to the data directory
    source : str
        The source of the data. Can be 'all', 'log', or 'net'

    Returns
    -------
    df : pd.DataFrame
        Dataframe containing the redwood data
    '''
    assert source in ['all', 'log', 'net'], 'source must be one of "all", "log", or "net"'
    redwood_data = pd.read_csv(oj(path, f'sonoma-data-{source}.csv'))
    redwood_data.columns = redwood_data.columns.str.strip()  # Remove any leading/trailing spaces
    return redwood_data


def load_mote_location_data(path='data'):
    ''' 
    Load the mote location data

    Parameters
    ----------
    path : str
        The path to the data directory

    Returns
    -------
    df : pd.DataFrame
        Dataframe containing the mote location data
    '''
    # TODO: LOAD MOTE LOCATION DATA
    return

File: labs/lab1/R/load.R
#' Load dates data
#'
#' @param path the path to the dates data files
#'
#' @returns A data frame containing the epoch information with three columns:
#'   epoch number, date, and day
load_dates_data <- function(path = here::here("data")) {
  epoch_nums <- read.table(
    file.path(path, "sonoma-dates-epochNums.txt"),
    col.names = "number"
  )
  epoch_dates <- read.table(
    file.path(path, "sonoma-dates-epochDates.txt"),
    col.names = "date"
  )
  epoch_days <- read.table(
    file.path(path, "sonoma-dates-epochDays.txt"),
    col.names = "day"
  )

  epoch_data <- cbind(epoch_nums, epoch_dates, epoch_days) |>
    tibble::as_tibble()
  return(epoch_data)
}


#' Load redwood data
#'
#' @param path the path to the redwood data files
#' @param source the source of the data to load: "all", "log", or "net"
#'
#' @returns A data frame containing the specified redwood data
load_redwood_data <- function(path = here::here("data"),
                              source = c("all", "log", "net")) {
  source <- match.arg(source)
  redwood_data <- data.table::fread(
    file.path(path, sprintf("sonoma-data-%s.csv", source))
  ) |>
    tibble::as_tibble()
  return(redwood_data)
}


#' Load mote location data
#'
#' @param path the path to the mote location data file
#'
#' @returns A data frame containing the mote location data
load_mote_location_data <- function(path = here::here("data")) {
  # TODO: LOAD MOTE LOCATION DATA
}
